### PR TITLE
[CPDLP-2965] Add NPQ application endpoints swagger docs

### PR DESCRIPTION
**Pull request by leandroalemao**
### Context

Ticket: [CPDLP-2965](https://dfedigital.atlassian.net/browse/CPDLP-2965)

We need to update the swagger documentation to include NPQ applications endpoint in all versions

### Changes proposed in this pull request

Add NPQ application endpoints swagger docs for all versions.

API swagger docs versions can be accessed via:
- [v1](https://npq-registration-review-1307-web.test.teacherservices.cloud/api/docs/v1)
- [v2](https://npq-registration-review-1307-web.test.teacherservices.cloud/api/docs/v2)
- [v3](https://npq-registration-review-1307-web.test.teacherservices.cloud/api/docs/v3)

[CPDLP-2965]: https://dfedigital.atlassian.net/browse/CPDLP-2965?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by ethax-ross — COMMENTED** [review]
If you branch off #1306  `2964-add-applications-index-endpoint` it will fill in the index/show actions

File: spec/requests/api/docs/v1/applications_spec.rb [hunk 1]
```diff
@@ -0,0 +1,83 @@
+require "rails_helper"
+require "swagger_helper"
+
+RSpec.describe "NPQ Applications endpoint", type: :request, openapi_spec: "v1/swagger.yaml" do
+  include_context "with authorization for api doc request"
+
+  let!(:application) { create(:application, lead_provider:) }
+
+  path "api/v1/npq-applications" do
```
Comment (ethax-ross) [inline spec/requests/api/docs/v1/applications_spec.rb:9]: Is it worth pulling a lot of this repeated logic that's going to be common to all index/show actions into `shared_contexts`? I did this with the request specs for applications/statements here and I think it makes it a lot cleaner/easier to maintain:

https://github.com/DFE-Digital/npq-registration/blob/1747c6826b1634e6338f2c115f09064a87cfd8e9/spec/support/shared_examples/api_show_endpoint_support.rb

https://github.com/DFE-Digital/npq-registration/blob/1747c6826b1634e6338f2c115f09064a87cfd8e9/spec/support/shared_examples/api_index_endpoint_support.rb

The specs then become:

https://github.com/DFE-Digital/npq-registration/blob/1747c6826b1634e6338f2c115f09064a87cfd8e9/spec/requests/api/v3/applications_spec.rb#L9-L33

File: spec/support/shared_examples/api_index_endpoint_documentation_support.rb [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+      tags tag
+      consumes "application/json"
+      produces "application/json"
+      security [api_key: []]
+
+      parameter name: :filter,
+                in: :query,
+                required: false,
+                schema: {
+                  "$ref": "#/components/schemas/ListApplicationsFilter",
```
Comment (ethax-ross) [inline spec/support/shared_examples/api_index_endpoint_documentation_support.rb:15]: The filter needs to be passed in as a parameter

File: spec/swagger_schemas/filters/list_applications.rb [hunk 3]
```diff
@@ -0,0 +1,55 @@
+LIST_APPLICATIONS_FILTER = {
+  v1: {
```
Comment (ethax-ross) [inline spec/swagger_schemas/filters/list_applications.rb:2]: As these schema versions are all very similar between versions could we construct the hash in a way that DRYs them up and makes it clear what the differences are between versions? Something like:

```
LIST_APPLICATIONS_FILTER = {
  v1: {
    ...
  }
}.tap do |h|
  h[v2] =  h[v1].merge({ 
    properties: { cohort: { description: "Overrides description" } 
  })
end.freeze
```

**Review by ethax-ross — COMMENTED** [review]
Looks good, just a couple other suggestions

File: spec/swagger_schemas/filters/list_applications.rb [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+}.tap { |h|
+  h[:v2] = h[:v1]
+  h[:v3] = h[:v1].merge({
+    properties: {
+      cohort: {
+        description: "Return only NPQ applications from the specified cohort or cohorts. This is a comma delimited string of years.",
+        type: "string",
+        example: "2021,2022",
+      },
+      updated_since: {
```
[the remark this answers is not in the document]
Comment (leandroalemao) [reply spec/swagger_schemas/filters/list_applications.rb:27]: Hi @ethax-ross tried to omit it but the order of filters will be different in v3 in comparison with v1/v2.. so left as it's..

**Omitting**
v1/v2:
- cohort
- updated_since

v3:
- cohort
- participant_id
- updated_since

**Not omitting**
v1/v2:
- cohort
- updated_since

v3:
- cohort
- updated_since
- participant_id

File: app/serializers/api/application_serializer.rb [hunk 5]
```diff
@@ -40,8 +40,8 @@ class AttributesSerializer < Blueprinter::Base
       end
 
       view :v3 do
-        # When we migrate schedules we can implement this fully.
-        field(:schedule_identifier) { "placeholder-schedule-identifier" }
+        # FIXME: When we migrate schedules we can test this fully.
```
Comment (cwrw) [inline app/serializers/api/application_serializer.rb:43]: 👍

File: spec/swagger_schemas/concerns/sorting.rb [hunk 6]
```diff
@@ -0,0 +1,4 @@
+SORTING_OPTIONS = {
+  description: "Sort records being returned.",
```
Comment (cwrw) [inline spec/swagger_schemas/concerns/sorting.rb:2]: could we add the list of accepted sorting params: 
```
Possible values:

created_at
-created_at
updated_at
-updated_at
```

File: spec/swagger_schemas/models/application.rb [hunk 7]
```diff
@@ -0,0 +1,445 @@
+APPLICATION = {
+  v1: {
```
Comment (cwrw) [inline spec/swagger_schemas/models/application.rb:2]: could we revisit nullable vs required fields to match the current schemas, example `course_identifier` will never be null etc.

**Review by cwrw — COMMENTED** [review]
looking good, thanks @leandroalemao. I like the way the schemas decide how we set the versions, however I think something we need to think about is how the schemas appear in different versions, for instance in v1: https://npq-registration-review-1307-web.test.teacherservices.cloud/api/docs/v1#/ I don't expect to see statements there or sorting etc. happy to chat through how we hide those and only show in relevant schemas

File: spec/swagger_helper.rb [hunk 8]
```diff
@@ -66,6 +66,9 @@
           SortingOptions: SORTING_OPTIONS,
         )
       end
+      if Rails.env.in?(%w[review])
```
Comment (cwrw) [inline spec/swagger_helper.rb:69]: Not sure this worked 🤔

**Review by cwrw — APPROVED** [review]
thanks @leandroalemao only comment around review maybe we can leave that for later to be generated at a different point as not sure that worked, otherwise 👍

File: spec/swagger_helper.rb (hunk 8 above)
Comment (leandroalemao) [reply spec/swagger_helper.rb]: Hi @cwrw It didn't work.. the idea was to add the review app url into the list of servers. Just removed. tks

**Review by cwrw — APPROVED** [review]

**Pull request merged**